[2024Q4] Decision Making Tool (#1714)
rolling-scopes-school/tasks: stage2/tasks/decision-making-tool/README.md

- Create a Pull Request from the development branch into the `main` branch and format it according to [PR requirements](https://docs.rs.school/#/pull-request-review-process?id=Требования-к-pull-request-pr). **Do not merge** the development branch into the `main` branch.
- Use the **PR link** to [submit](https://app.rs.school/course/student/dashboard) the app to mentor in [app.rs](https://app.rs.school/course/student/dashboard).

## General Requirements

- `<body>` in the `index.html` must be empty (only `<script>` tag is allowed). _(You can check this with the `View page source` option in the context menu in Google Chrome.)_
- App must be supported at `640px <= width <= 1280px` at least _(DOM elements should not go out of bounds of parent elements, overlap, etc.)_.
- The use of `alert`, `prompt`, `confirm` is prohibited.
- App must not have unexpected errors in the console.
- The app must be supported by at least the latest version of the Google Chrome browser.

## Technical Requirements (160 points total)

### General

- Language: Application written in TypeScript.

> - Required: Prettier, ESLint, StyleLint and libraries for them.

Author: ```suggestion
- Required: Prettier, ESLint, StyleLint. _(It is allowed to use it's configs and plugins.)_
```

> - Required: Prettier, ESLint, StyleLint and libraries for them.

Reviewer: 👌

- Required: Husky, lint-staged, commitlint.
- Required: Bundlers (webpack, vite). _(It is allowed to use bundler's plugins.)_
- Allowed: CSS modules, CSS Preprocessors (`Sass`, `Less`, `Stylus`, `PostCSS`), CSS in JS libraries (`tailwindcss`, `styled components`), `clsx`/`classnames` package.
- Prohibited: Frameworks like Angular, React, Vue, etc.
- Prohibited: jQuery.
- Prohibited: eslint-plugin-prettier.
- Prohibited: Third-party libraries not listed in the allowed libraries.

### Code Formatting and Linting (80 points)

#### TypeScript (12/80)

> TypeScript helps us avoid unnecessary code errors at runtime and helps the IDE give us relevant hints during development.

1. (+2) [TypeScript](https://www.typescriptlang.org/) must be installed to developer dependencies.
2. (+2) TypeScript configuration file is added to the project and used.

> 3. (+8) TypeScript configuration file must include the `strict` options enabled.

Reviewer: Please, mention this explicitly

> 3. (+8) TypeScript configuration file must include the `strict` options enabled.

Author: ```suggestion
3. (+8) TypeScript configuration file must have all of the [strict mode family options](https://www.typescriptlang.org/tsconfig/#strict) enabled.
```

#### Prettier (8/80)

> The formatter allows us to be less distracted by manual indentation and comma spacing, allowing us to better focus on writing code.

1. (+2) [Prettier](https://prettier.io/) must be installed to developer dependencies.
2. (+2) Prettier configuration file is added to the project and used.

> 3. (+4) `package.json` must contain 2 scripts: `ci:format` for checking issues with CI and `format` for manually formatting project files. _The `prettier` command has many useful options, such as `--list-different` and `--ignore-unknown`._

Author: Good point.
I think we should add somithing like

> Briefly, CI (Continuous Integration) is the process of automatically checking code against specified standards.
> 
> `ci:` scripts (e.g., `ci:format`, `ci:lint`) are commands that check code but do not correct errors (used for validation in CI systems or locally).

under the heading of the “Code Formatting and Linting” section, or under the heading of the `husky` section (but then it would be better to elevate the `husky` section to the first or second place).

The idea with such `ci:` scripts I took from the repositories of rolling scopes school ([rsapp](https://github.com/rolling-scopes/rsschool-app/blob/master/package.json), [tasks](https://github.com/rolling-scopes-school/tasks/blob/master/package.json)).

> 3. (+4) `package.json` must contain 2 scripts: `ci:format` for checking issues with CI and `format` for manually formatting project files. _The `prettier` command has many useful options, such as `--list-different` and `--ignore-unknown`._

Reviewer: In such case you should add such "small tips" for every instrument you mention here for consistency 😉

> 3. (+4) `package.json` must contain 2 scripts: `ci:format` for checking issues with CI and `format` for manually formatting project files. _The `prettier` command has many useful options, such as `--list-different` and `--ignore-unknown`._

Reviewer: Yeah, but those scripts are executed on CI, not locally. Husky is not a CI. CI is not about checking code against specified standards.
If we take your example, there only diff in scripts is about `--write` and `--check` flags. But when we run commands locally we want exectly `--write`, not `--check`.

> 3. (+4) `package.json` must contain 2 scripts: `ci:format` for checking issues with CI and `format` for manually formatting project files. _The `prettier` command has many useful options, such as `--list-different` and `--ignore-unknown`._

Author: Almost all italicized phrases are either clarifications or such tips. They are present throughout the file, not just in the "Prettier" block 😄

> 3. (+4) `package.json` must contain 2 scripts: `ci:format` for checking issues with CI and `format` for manually formatting project files. _The `prettier` command has many useful options, such as `--list-different` and `--ignore-unknown`._

Reviewer: I don't think we need any prefix at all. 
The one shouldn't run all those scripts manually, as he could use actions on save.
 If we have any commands we need only for lint-staged, we don't need them in `scripts`.
If you want different scritps for prettier you could use `format: write` and `format: check`.

#### ESLint (40/80)

> Linter allows us to keep our code clean. Clean code is more readable, maintainable, and reusable. And its auto-fixing of errors allows us to save time on refactoring.

1. (+2) [ESLint](https://eslint.org/) must be installed to developer dependencies.
2. (+2) ESLint configuration file is added to the project and used. _(both [`flat`](https://eslint.org/docs/latest/use/configure/configuration-files) and [`legacy`](https://eslint.org/docs/latest/use/configure/configuration-files-deprecated) config formats are allowed.)_

> 3. (+8) ESLint configuration file must include the `noInlineConfig` and `reportUnusedDisableDirectives` options enabled.

Reviewer: 👌

> 4. (+8) ESLint configuration file must include all of the following rules included:

Reviewer: 👌

     "@typescript-eslint/consistent-type-assertions": [
       "error",
       { "assertionStyle": "never" }
     ],
     "@typescript-eslint/explicit-member-accessibility": [
       "error",
       { "accessibility": "explicit", "overrides": { "constructors": "off" } }
     ],
     "@typescript-eslint/member-ordering": "error"
   }
   ```

5. (+16) ESLint configuration file must be configured with [`typescript-eslint`](https://typescript-eslint.io/) (with enabled [type checking rules](https://typescript-eslint.io/getting-started/typed-linting/)), [`eslint-config-airbnb-typescript`](https://www.npmjs.com/package/eslint-config-airbnb-typescript), [`eslint-config-prettier`](https://github.com/prettier/eslint-config-prettier) (do not confuse it with `eslint-plugin-prettier`. it is [not recommended](https://prettier.io/docs/en/integrating-with-linters.html#:~:text=generally%20not%20recommended) by the prettier documentation.) and the necessary dependencies for them to work.
   - _With the mentor's permission, it is acceptable to disable or tweak some rules (e.g. `import/prefer-default-export`, `no-underscore-dangle`, etc.) as long as it does not conflict with penalties._

>    - _It is allowed to tweak the config to be stricter to make your code even cleaner (according to your or your mentor's preference)._

Author: Here's the thing:
`eslint-config-airbnb-base` and `eslint-config-airbnb-typescript` look outdated at the moment (have not been updated to ESLint v9). 
A possible solution is to remove airbnb from the requirements and add a list of important good eslint rules, but I need help with a list of such rules.

Reviewer: I could suggest https://www.npmjs.com/package/eslint-plugin-unicorn.
Typescript-eslint itself uses it https://github.com/typescript-eslint/typescript-eslint/blob/main/eslint.config.mjs

3. (+2) `husky` must be configured to run `commitlint` on the `commit-msg` hook.
4. (+2) `lint-staged` must be configured to run `ci:` scripts at least for javascript/typescript and css/scss/etc. files.
5. (+2) `commitlint` must be configured with [`@commitlint/config-conventional`](https://www.npmjs.com/package/@commitlint/config-conventional).

### Code Quality (80 points)

> The use of innerHTML is prohibited. Use of DOM search methods is prohibited. ([there are good reasons for not using them](https://gist.github.com/TELEUZI/410d19772481d98b06e0b41ebf89fff1#naive-implementation-).)
>
> For creating elements that will need to be accessed by code, a great solution is to use `createElement` and variables. For clearing/replacing element content, `replaceChildren` may be useful. To add elements that will not need to be accessed by code, it is acceptable to use `insertAdjacentHTML`.
>
> This won't cause any pain if you at least just write a reusable function that takes in the necessary element settings (`tagName`, `className`, `textContent`, etc.) and returns a ready-to-use element.

#### Code Smells (4/80)

> 1. (+2) The code does not contain magic values.

Reviewer: I mean we should add this info for students. 
"magic values" concept isn't a common knowledge among the students I think.

> 1. (+2) The code does not contain magic values.

Author: ```suggestion
1. (+2) The code does not contain magic (hardcoded) values. _([magic numbers](https://refactoring.guru/replace-magic-number-with-symbolic-constant), [magic strings](https://dev.to/josuews303/dont-use-magic-strings-2aol), etc.)_
```

2. (+2) Code contains minimal or no code duplication at all.

#### Bundling and Modularity (12/80)

1. (+4) HTML elements are generated using a utility function or class.
2. (+4) The code is logically divided into modules/layers. _(you can consult your mentor about architecture.)_
3. (+4) The application is built by a bundler.

#### Clear and Reusable Functions (16/80)

> 1. (+8) The code is divided into small functions (≤ 40 lines each) with clear names and purposes. The same applies to classes and their methods. _(ESLint rule `max-lines-per-function` may be useful for this.)_

Author: ```suggestion
1. (+8) The code is divided into small functions (≤ 40 lines each or according to the mentor's preference) with clear names and purposes. The same applies to classes and their methods. _(ESLint rule `max-lines-per-function` may be useful for this.)_
```

2. (+8) Arguments and return values of functions and methods are explicitly typed.

#### Care About Types (48/80)

> [There is no point in using TypeScript if you don’t care about types](https://javascript.plainenglish.io/there-is-no-point-to-use-typescript-in-your-project-if-you-dont-care-about-types-68131deeb43a)

1. (+16) The code does not contain any type assertions.
2. (+16) The code does not contain any explicit or implicit `any`.
3. (+16) The code contains and uses [type guards](https://www.typescriptlang.org/docs/handbook/2/narrowing.html) _(e.g., type narrowing type predicates, assertion functions, etc.)_. _(small hint: functions to get data from a json file and/or localStorage are great places to apply type guards.)_

> ## Functional Requirements (256 points total)

Reviewer: Could we make it 415 too? 🙂

> ## Functional Requirements (256 points total)

Author: ```suggestion
## Functional Requirements (255 points total)
```
